packages/dds/tree/src/test/shared-tree/sharedTree.bench.ts

```diff
@@ -0,0 +1,877 @@
+			tree[0].value = 123;
+			break;
+		case TreeShape.Wide:
+			nodesUnderRoot = tree.fields.foo.length;
+			if (nodesUnderRoot === 0) {
+				tree.fields.value = 123;
+			} else {
+				tree.fields.foo[nodesUnderRoot - 1].value = 123;
+			}
+			break;
+		default:
+			unreachableCase(shape);
+	}
+}
+
+function getLeafNodeFromJSObject(tree: Jsonable): Jsonable {
+	for (const key of Object.keys(tree)) {
+		if (typeof tree[key] === "object" && tree[key] !== null) {
+			if (tree[key].type !== undefined && tree[key].fields === undefined) {
+				return tree;
+			}
+			return getLeafNodeFromJSObject(tree[key]);
+		}
+	}
+}
+
+function readCursorTree(forest: IForestSubscription, numberOfNodes: number, shape: TreeShape) {
+	const readCursor = forest.allocateCursor();
+	moveToDetachedField(forest, readCursor);
+	assert(readCursor.firstNode());
+	switch (shape) {
+		case TreeShape.Deep:
+			for (let i = 0; i < numberOfNodes; i++) {
+				readCursor.firstField();
+				assert(readCursor.firstNode());
+			}
+			break;
+		case TreeShape.Wide:
+			if (numberOfNodes === 1) {
+				break;
+			}
+			readCursor.firstField();
+			readCursor.firstNode();
+			for (let j = 0; j < numberOfNodes; j++) {
+				readCursor.nextNode();
+			}
+			break;
+		default:
+			unreachableCase(shape);
+	}
+	readCursor.free();
+}
+
+/**
+ * Given a tree and path, changes the node value to a different one.
+ * @param tree - tree that you need to manipulate
+ * @param path - location where you need to apply the edit
+ * @param dataType - the primitive data type of the value to store
+ */
+function manipulateCursorTree(tree: ISharedTree, path: UpPath, dataType: TestPrimitives) {
+	const value = generateTreeData(dataType);
+	tree.runTransaction((forest, editor) => {
+		editor.setValue(path, { type: brand("Test"), value });
+		return TransactionResult.Apply;
+	});
+}
+
+function getCursorLeafNode(numberOfNodes: number, shape: TreeShape): UpPath {
+	let path: UpPath;
+	switch (shape) {
+		case TreeShape.Deep:
+			path = {
+				parent: undefined,
+				parentField: rootFieldKeySymbol,
+				parentIndex: 0,
+			};
+			for (let i = 0; i < numberOfNodes; i++) {
+				path = {
+					parent: path,
+					parentField: localFieldKey,
+					parentIndex: 0,
+				};
+			}
+			return path;
+		case TreeShape.Wide:
+			path = {
+				parent: {
+					parent: undefined,
+					parentField: rootFieldKeySymbol,
+					parentIndex: 0,
+				},
+				parentField: localFieldKey,
+				parentIndex: numberOfNodes - 1,
+			};
+			return path;
+		default:
+			unreachableCase(shape);
+	}
+}
+
+function readEditableTree(tree: ISharedTree, numberOfNodes: number, shape: TreeShape) {
+	assert(isUnwrappedNode(tree.root));
+	let currField;
+	let currNode;
+	switch (shape) {
+		case TreeShape.Deep:
+			currField = tree.root[getField](localFieldKey);
+			currNode = currField.getNode(0);
+			for (let j = 0; j < numberOfNodes - 1; j++) {
+				currField = currNode[getField](localFieldKey);
+				currNode = currField.getNode(0);
```

Comment:
assuming your schema is using a value field, I think a more conventional way to do this would be 
```suggestion
				currNode = currField.foo;
```

or:
```suggestion
				currNode = currField[localFieldKey];
```

Reply:
If I delete the `currField` declaration, the `currNode = currField[localFieldKey]` line does not work as `currField` is no longer defined. Should it be changed to the following instead?
```
currField = currNode[getField](localFieldKey);
currNode = currField[localFieldKey];
```

Reply:
You should not need access to the field  objects in that case. Something like:

```
function readEditableTree(tree: ISharedTree, numberOfNodes: number, shape: TreeShape) {
	switch (shape) {
		case TreeShape.Deep: {
			let currentNode: UnwrappedEditableField = tree.root;
			for (let j = 0; j < numberOfNodes - 1; j++) {
				assert(isUnwrappedNode(currentNode));
				currentNode = currentNode.foo as UnwrappedEditableField;
			}
			break;
		}
		case TreeShape.Wide: {
			const root = tree.root;
			assert(isUnwrappedNode(root));
			const field = root.foo as UnwrappedEditableField;
			assert(isEditableField(field));
			for (const iterator of field) {
			}
			break;
		}
		default:
			unreachableCase(shape);
	}
}
```

Also note that I replaced "curr" with "current": its best to use full English words in names.

Also with benchmarks is generally good practice to do something with the data you are supposed to be accessing to help ensure the code is more realistic and the optimizer can't optimize it in atypical ways. One common think we do in our benchmarks similar to this is is check each node value for a number and add them all up. We even already have functions to do this using cursors ( src/test/domains/json/benchmarks.ts )

Reply:
The type casting is my example is bad, but tats an issue with the current version of EditableTree. Don't go doing things like that elsewhere

Reply:
Updated readEditableTree function